File: renewal_module/renewal_module/report/customers_data/customers_data.py
# Copyright (c) 2025, Aravind Mandala and contributors
# For license information, please see license.txt

# import frappe
# def execute(filters=None):
# 	columns, data = [], []
# 	return columns, data

import frappe

def execute(filters=None):
    

    based_on = filters.get("based_on", "Customers")

    if based_on == "Customers":
        columns, data = get_customers_data(filters)
    elif based_on == "Customers Without Contacts":
        columns, data = get_customers_without_contacts(filters)
    elif based_on == "Customers With Contacts And No Opps":
        columns, data = get_customers_with_contacts_no_opps(filters)
    elif based_on == "Customers With Contacts With Opps":
        columns, data = get_customers_with_contacts_with_opps(filters)
    elif based_on == "Customers With Contacts And With Opps And No Sales Invoice":
        columns, data = get_customers_with_contacts_opps_no_invoice(filters)
    elif based_on == "Customers Without Renewals":
        columns, data = get_customers_without_renewals(filters)
    else:
        frappe.throw(f"Unknown filter option: {based_on}")



    currency = filters.presentation_currency or frappe.get_cached_value(
        "Company", filters.company, "default_currency"
    )

    # Initialize select_row to 0 for all rows
    for row in data:
        row["select_row"] = 0
        
    report_summary = get_report_summary(filters,columns, currency, data)
    # frappe.msgprint("<pre>{}</pre>".format(frappe.as_json(report_summary)))
    chart = get_chart_data(filters, columns, data)
    # frappe.msgprint("<pre>{}</pre>".format(frappe.as_json(chart)))

    return columns, data,None, chart, report_summary
    #return columns,data


def get_customers_data(filters):
    conditions, values = get_conditions(filters)
    where_clause = f"1=1 {conditions}"
    data = frappe.db.sql(f"""
        SELECT 0 AS select_row, tc.name as customer, tc.owner as owner, tst.sales_person
        FROM tabCustomer tc
        INNER JOIN `tabSales Team` tst ON tst.parent = tc.name
        WHERE {where_clause}
        GROUP BY tc.name
        ORDER BY tc.name
    """, values, as_dict=True)

    return get_columns(), data




def get_customers_without_contacts(filters):
    conditions, values = get_conditions(filters)
    data = frappe.db.sql(f"""
        SELECT 0 AS select_row, tc.name as customer, tc.owner as owner, tst.sales_person
        FROM tabCustomer tc
        INNER JOIN `tabSales Team` tst ON tst.parent = tc.name
        WHERE NOT EXISTS (
            SELECT 1 FROM `tabDynamic Link` tdl
            JOIN tabContact c ON tdl.parent = c.name
            WHERE tdl.link_doctype = 'Customer'
            AND tdl.link_name = tc.name
        ) {conditions}
        GROUP BY tc.name
    """, values, as_dict=True)

    return get_columns(), data


def get_customers_with_contacts_no_opps(filters):
    conditions, values = get_conditions(filters)
    data = frappe.db.sql(f"""
        SELECT DISTINCT tc.name AS customer, tc.owner, tst.sales_person
        FROM tabCustomer tc
        INNER JOIN `tabSales Team` tst ON tst.parent = tc.name
        WHERE EXISTS (
            SELECT 1 FROM `tabDynamic Link` tdl
            JOIN tabContact c ON tdl.parent = c.name
            WHERE tdl.link_doctype = 'Customer' AND tdl.link_name = tc.name
        )
        AND NOT EXISTS (
            SELECT 1 FROM `tabOpportunity` o WHERE o.party_name = tc.name
        ) {conditions}
        GROUP BY tc.name
    """, values, as_dict=True)

    return get_columns(), data


def get_customers_with_contacts_with_opps(filters):
    conditions, values = get_conditions(filters)
    data = frappe.db.sql(f"""
        SELECT DISTINCT tc.name AS customer, tc.owner, tst.sales_person
        FROM tabCustomer tc
        INNER JOIN `tabSales Team` tst ON tst.parent = tc.name
        WHERE EXISTS (
            SELECT 1 FROM `tabDynamic Link` tdl
            JOIN tabContact c ON tdl.parent = c.name
            WHERE tdl.link_doctype = 'Customer' AND tdl.link_name = tc.name
        )
        AND EXISTS (
            SELECT 1 FROM `tabOpportunity` o WHERE o.party_name = tc.name
        ) {conditions}
        GROUP BY tc.name
    """, values, as_dict=True)

    return get_columns(), data



def get_customers_with_contacts_opps_no_invoice(filters):
    conditions, values = get_conditions(filters)
    data = frappe.db.sql(f"""
        SELECT DISTINCT tc.name AS customer, tc.owner, tst.sales_person
        FROM tabCustomer tc
        INNER JOIN `tabSales Team` tst ON tst.parent = tc.name
        WHERE EXISTS (
            SELECT 1 FROM `tabDynamic Link` tdl
            JOIN tabContact c ON tdl.parent = c.name
            WHERE tdl.link_doctype = 'Customer' AND tdl.link_name = tc.name
        )
        AND EXISTS (
            SELECT 1 FROM `tabOpportunity` o WHERE o.party_name = tc.name
        )
        AND NOT EXISTS (
            SELECT 1 FROM `tabSales Invoice` si WHERE si.customer = tc.name
        ) {conditions}
        GROUP BY tc.name
    """, values, as_dict=True)

    return get_columns(), data



def get_customers_without_renewals(filters):
    conditions, values = get_conditions(filters)
    data = frappe.db.sql(f"""
        SELECT DISTINCT tc.name AS customer, tc.owner, tst.sales_person
        FROM tabCustomer tc
        INNER JOIN `tabSales Team` tst ON tst.parent = tc.name
        WHERE NOT EXISTS (
            SELECT 1 FROM `tabRenewal List` r WHERE r.customer_name = tc.name
        ) {conditions}
        GROUP BY tc.name
    """, values, as_dict=True)

    return get_columns(), data


def get_columns():
    return [
        {"label": "Select", "fieldname": "select_row", "fieldtype": "Check", "width": 60},
        {"label": "Customer", "fieldname": "customer", "fieldtype": "Link", "options": "Customer", "width": 300},
        {"label": "Sales Person", "fieldname": "sales_person", "fieldtype": "Data", "width": 200},
        {"label": "Created By", "fieldname": "owner", "fieldtype": "Data", "width": 200}
    ]

def get_allowed_sales_persons(user):
    if user == "Administrator" or "System Manager" in frappe.get_roles(user):
        return None  # Means unrestricted access (show all)
    employee = frappe.db.get_value("Employee", {"user_id": user}, "name")
    if not employee:
        return []
    sales_person_doc = frappe.get_all("Sales Person", filters={"employee": employee}, fields=["name", "is_group"])
    if not sales_person_doc:
        return []
    sales_person = sales_person_doc[0]["name"]
    is_group = sales_person_doc[0]["is_group"]
    if is_group:
        children = frappe.get_all("Sales Person", filters={"parent_sales_person": sales_person}, pluck="name")
        return [sales_person] + children
    else:
        return [sales_person]



def get_conditions(filters):
    conditions = []
    values = {}

    user = frappe.session.user
    selected_sales_persons = filters.get("sales_person")

    if not selected_sales_persons:
        allowed_salespersons = get_allowed_sales_persons(user)
        if allowed_salespersons is not None:
            conditions.append("tst.sales_person IN %(allowed_sales_persons)s")
            values["allowed_sales_persons"] = tuple(allowed_salespersons)
    else:
        conditions.append("tst.sales_person IN %(sales_person)s")
        values["sales_person"] = tuple(selected_sales_persons)
	
    # Sales Person: MultiSelectList
    if filters.get("sales_person"):
        conditions.append("tst.sales_person IN %(sales_person)s")
        values["sales_person"] = tuple(filters.get("sales_person"))
    # Sales Person: MultiSelectList
    # if filters.get("sales_person"):
    #     conditions.append("tst.sales_person IN %(sales_person)s")
    #     values["sales_person"] = tuple(filters.get("sales_person"))

    if filters.get("territory"):
        conditions.append("tc.territory IN %(territory)s")
        values["territory"] = tuple(filters.get("territory"))	

    if filters.get("industry"):
        conditions.append("tc.industry IN %(industry)s")
        values["industry"] = tuple(filters.get("industry"))	    

    # Created By (User): MultiSelectList
    if filters.get("created_by"):
        conditions.append("tc.owner IN %(created_by)s")
        values["created_by"] = tuple(filters.get("created_by"))

    # Starts With (Customer Name)
    if filters.get("starts_with"):
        letters = filters.get("starts_with")
        like_conditions = []
        for i, letter in enumerate(letters):
            key = f"starts_with_{i}"
            like_conditions.append(f"tc.name LIKE %({key})s")
            values[key] = f"{letter}%"
        if like_conditions:
            conditions.append("(" + " OR ".join(like_conditions) + ")")

    return (" AND " + " AND ".join(conditions)) if conditions else "", values



def get_report_summary(filters,columns, currency, data):
    customer_count = 0
    
    # frappe.msgprint("<pre>{}</pre>".format(frappe.as_json(new)))

    if data:
        for period in data:
            if period.customer :
                customer_count += 1		

    customer_label = ("Customer Count")

    return [
        {
            "value": f"{customer_count:,}",
            "indicator": "Green",
            "label": "Customer Count",
            "datatype": "Data"
        }
    ]


def get_chart_data(filters, columns, data):
    customer_count = 0
    if data:
        for period in data:
            if period.customer:
                customer_count += 1

    return {
        "data": {
            "labels": ["Customers Count"],
            "datasets": [{"values": [customer_count]}],
        },
        "type": "pie",
        "colors": ["#c80064"],
    }

File: renewal_module/renewal_module/report/customers_data/customers_data.js
// Copyright (c) 2025, Aravind Mandala and contributors
// For license information, please see license.txt

frappe.query_reports["Customers Data"] = {
	"filters": [

	]
};

frappe.query_reports["Customers Data"] = {
	"filters": [

		// {
		// 	"fieldname": "party_name",
		// 	"label": __("Party"),
		// 	"fieldtype": "MultiSelectList",
		// 	"options": "Customer",
		// 	get_data: function (txt) {
		// 		return frappe.db.get_link_options('Customer', txt);
		// 	},

		// },
		{
			fieldname: "sales_person",
			label: __("Sales Person"),
			fieldtype: "MultiSelectList",
			default: [],
			get_data: function (txt) {
				return frappe.call({
					method: "renewal_module.api.get_salespersons_for_user"
				}).then(r => {
					const results = (r.message || []).filter(item =>
						item.name.toLowerCase().includes(txt.toLowerCase())
					);
					return results.map(item => ({
						value: item.name,
						description: item.parent_sales_person
					}));
				});
			}

		},

		{
			"fieldname": "territory",
			"label": __("Territory"),
			"fieldtype": "MultiSelectList",
			"options": "Territory",
			get_data: function (txt) {
				return frappe.db.get_link_options('Territory', txt);
			},
		},
		{
			"fieldname": "industry",
			"label": __("Industry Type"),
			"fieldtype": "MultiSelectList",
			"options": "Industry",
			get_data: function (txt) {
				return frappe.db.get_link_options('Industry Type', txt);
			},
		},
		{
			"fieldname": "created_by",
			"label": __("Created By"),
			"fieldtype": "MultiSelectList",
			"options": "User",
			get_data: function (txt) {
				return frappe.db.get_link_options('User', txt);
			}
		},
		{
			fieldname: "starts_with",
			label: __("Starts With"),
			fieldtype: "MultiSelectList",
			options: ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"],
			multiple: 1
		},
		{

			"fieldname": "based_on",
			"label": __("Based On"),
			"fieldtype": "Select",
			"options": "Customers\nCustomers Without Contacts\nCustomers With Contacts And No Opps\nCustomers With Contacts With Opps\nCustomers With Contacts And With Opps And No Sales Invoice\nCustomers Without Renewals",
			"default": "Customers"
		}


	],

	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		if (column.fieldname === "select_row") {
			const checked = data.select_row ? 'checked' : '';
			const customer_id = frappe.utils.escape_html(data.customer || "");
			return `<input type="checkbox" class="report-checkbox" data-row='${customer_id}' ${checked}>`;
		}
		return value;
	},

	onload: function (report) {
		// Check current roles before setting default
		const user = frappe.session.user;
		const roles = frappe.user_roles || [];

		// Only auto-set filter for non-admin, non-system manager
		if (user !== "Administrator" && !roles.includes("System Manager")) {
			frappe.call({
				method: "renewal_module.api.get_salespersons_for_user",
				callback: function (r) {
					const options = r.message || [];
					if (options.length > 0) {
						const firstSalesPerson = options[0].name;
						report.set_filter_value("sales_person", [firstSalesPerson]);
					}
				}
			});
		}

		// Add checkboxes to the report
		let assignBtn;

		function setCheckboxes(checked = true) {
			const dt = report.datatable;
			if (!dt || !dt.wrapper) {
				frappe.msgprint("No data available.");
				return;
			}
			console.log(dt)

			// Update internal row data
			if (dt.datamanager?.data) {
				dt.datamanager.data.forEach(row => {
					row.select_row = checked ? 1 : 0;
				});
			}

			// Refresh datatable to reflect checkbox state
			dt.refresh();

			updateAssignButtonVisibility();
		}

		function updateAssignButtonVisibility() {
			const anyChecked = !!document.querySelector('.report-checkbox:checked');
			if (assignBtn) {
				assignBtn.toggle(anyChecked);
			}
		}

		report.page.add_inner_button(__('Select All'), () => {
			setCheckboxes(true);
		});

		report.page.add_inner_button(__('Clear Selection'), () => {
			setCheckboxes(false);
		});

		assignBtn = report.page.add_inner_button("Assign User", function () {
			// let selected = [];
			// $('.report-checkbox:checked').each(function () {
			// 	selected.push($(this).attr("data-row"));
			// });
			let selected = [];
			const dt = report.datatable;
			if (dt?.datamanager?.data) {
				dt.datamanager.data.forEach(row => {
					if (row.select_row) {
						selected.push(row.customer);  // or whatever your unique key is
					}
				});
			}
			console.log("Selected customers:", selected);
			if (!selected.length) {
				frappe.msgprint("Please select at least one customer.");
				return;
			}

			frappe.prompt([
				{
					label: 'User Email',
					fieldname: 'user_email',
					fieldtype: 'Link',
					options: 'User',
					reqd: 1
				}
			], function (values) {
				frappe.call({
					method: 'renewal_module.api.bulk_assign_customer_to_salesperson_by_user',
					args: {
						customers: selected,
						user_email: values.user_email
					},
					callback: function (r) {
						if (r.message === "ok") {
							frappe.msgprint("User assigned successfully.");
							report.refresh();
						} else {
							frappe.msgprint("Something went wrong: " + (r.message || "No response."));
						}
					}
				});
			}, 'Assign User', 'Assign');
		});

		assignBtn.hide(); // initially hidden

		// Sync checkbox UI with datamanager data
		$(document).on('change', '.report-checkbox', function () {
			const rowKey = $(this).attr("data-row");
			const isChecked = $(this).is(":checked");

			const dt = report.datatable;
			if (dt?.datamanager?.data) {
				let row = dt.datamanager.data.find(r => r.customer === rowKey);
				if (row) {
					row.select_row = isChecked ? 1 : 0;
				}
			}

			updateAssignButtonVisibility();
		});
	}



};